### ISSUE TITLE
atoallinks.com domain page broke

### ISSUE THREAD

**Issue by tripleee** [Charcoal-SE/metasmoke#897]
I tried to add a link to sevenmentor.com to the domain page
https://metasmoke.erwaysoftware.com/domains/51526
but this seems to have broken it.

Is the record corrupted, or is there something else going on here?

Visiting the page should get you a Rails traceback with more information. It looks like the linked domain information is somehow missing or incorrect.

**Comment by makyen** [comment]
Yes, the entries which you created in the domain_links table didn't have a `right_id` (i.e. a domain to link).

| id    | left_id | right_id | link_type         | comments | creator_id | created_at               | updated_at              | 
|-------|---------|----------|-------------------|----------|------------|--------------------------|-------------------------| 
| 1982  | 51526   |          | Users associated  |          | 41         | 2022-01-04 10:29:45 UTC  | 2022-01-04 10:29:45 UTC | 
| 1983  | 51526   |          | Users associated  |          | 41         | 2022-01-04 10:29:45 UTC  | 2022-01-04 10:29:45 UTC | 

I determined this by using the Blazer/SQL Data Exploder to look for malformed entries in that table.

I've deleted those entries in the table by sending POSTs to `https://metasmoke.erwaysoftware.com/domains/links/1982` and `https://metasmoke.erwaysoftware.com/domains/links/1983` with `_method=delete`. So, the 51526 domain properly displays.

We should update MS such that A) malformed entries are not created, and B) the code to display them ignores any malformed entries.

**Comment by tripleee** [comment]
Thanks. Should I retitle this bug report or create a new one?

**Comment by makyen** [comment]
Either way works. If you create a new issue, then just link to this one for a bit of context.

**Comment by makyen** [comment]
Bookmakrlet which will prompt for domain link to delete:
```
javascript:void((function(){
    function deleteMSDomainLink(domainLinkId) {
        if (domainLinkId) {
            const authenticity_token = $(document.head).find('meta[name="csrf-token"]').attr('content').trim();
            if (!authenticity_token) {
                alert('ABORTING: Unable to determine authenticity_token');
                return;
            }
            const _method = 'delete';
            $.post(`https://metasmoke.erwaysoftware.com/domains/links/${domainLinkId}`, {authenticity_token, _method});
        }
    }
    domainLinkId = ((prompt('What is the ID of the domain link you want to delete?') || '').match(/\d+/) || [])[0];
    if (domainLinkId && confirm(`Are you sure you want to delete domain link ID: ${domainLinkId}`)) {
        deleteMSDomainLink(domainLinkId);
    }
})());
```

SQL Data Explorer query created by @tripleee to show the domain links with a specified domain ID: https://metasmoke.erwaysoftware.com/data/sql/queries/291-domain-links-on-domain-page

**Comment by tripleee** [comment]
Worked now, dunno what was different.

**Comment by tripleee** [comment]
... Oh, the bookmarklet only works when I open it on a Metasmoke page.

**Comment by makyen** [comment]
This issue should be substantially reduced, but not eliminated. We should still change the MS backend to validate that there's all of a `left_id`, `right_id`, and `link_type` for each new entry prior to creation.

It should, however, now be much harder to inadvertently create invalid entries through the MS user interface (i.e. the New Domain Link modal dialog). The combination of the commits [domain-link modal: domain and why are required](https://github.com/Charcoal-SE/metasmoke/commit/2f3cddad91841ef1417f938e69528b2636cf1613) and [typo: required isn't a data attribute](https://github.com/Charcoal-SE/metasmoke/commit/dd98c111688651d46e6f2a2cde04225bbef41cf5) make the `right_id` and `link_type` required in order to submit the new domain link modal dialog. That should make it substantially less likely mistakenly create invalid entries through the UI on MS.
